### PR TITLE
Econ 2.0 Buildings

### PR DESCRIPTION
**Pull request by gawquon**
---
### Demand Building Component
- [x] ~~Load employment from PMs~~ (Already did, thanks past me.)
- [x] ~~Load input/output goods from PMs~~
- [x] Use average weight of PMs from EconManager config file. (Country by country basis)
- [x] Predict ownership PMs based on country laws.
- [x] Track available workers and Subsistence space.
- [x] Track Jobs% and Culture% when building.
- [x] Resolves a building's PMGroups to building's average net input/output/jobs.
- [x] Add in supply created/lost by subsistence building.

### ~~Balancing~~ - Next PR
~~- [ ] Have demand be primary factor, but allow the other weights to punch through at times.~~
~~- [ ] Add function to generate extra arable land for extra pop.~~
~~- [ ] Check additional bureaucracy cost incurred by nationalization, supplementary admin round if needed.~~

### Jobs
- [x] Use working ratios to hire/fire the right number of workers + dependents
- [x] Initialize the starting jobs in a country substate by substate. Subsistence Buildings, Manor Houses, and unemployed.
- [x] Predict PM's based on laws. (Serfdom vs Homesteading, etc.)
- [x] Peg loss of peasants to subsequent job loss in Manor Houses (Send them to unemployed)

### Buildings
- [x] Track Economies of scale for everything except subsistence, capped by tech.
- [x] Load State modifiers from State traits. Match the building to a trait.
- [x] Match Output Goods to any state trait
- [x] Add Urban Centers alongside ownership calcs
- [x] Integrate Hardcoded buildings (Ports and Gov. Admins) into the market.
- [ ] (Optional): Use ownership estimations to modify the throughput Economies of scale

### Debug
- [x] Make a formatted table output for the Market
- [x] Make a formatted table output for Culture and Job maps.

### Ownership
- [x] Resolve ownership rules to estimates on the number of Manor Houses / Financial Districts

### CODE REVIEW

**Review by Zemurin — CHANGES_REQUESTED** [review]
Mostly semantics, looks very very good overall. I'm sure Idhrendur will find tons more though.

File: EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingResources.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	{
+		int index = 0;
+		double percent = 1.0;
+
+		// If a rule exists
+		if (const auto& estimateIter = estimatedPMs.find(PMGroupName); estimateIter != estimatedPMs.end())
+		{
+			const auto& [i, p] = estimateIter->second;
+			index = i;
+			percent = p;
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingResources.cpp:18]: I grant that is the most literal definition. In my experience, (I don't know if it's just an American thing or Mathematician), the working definition of when we use fraction is to refer to x/y notation for rationals, and we use percent to refer to any decimal notation whether that's 0.70 or 70%.

Regardless, I'll switch it.

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	for (const auto& [type, percent]: estimatedOwnerships)
+	{
+		for (const auto& [job, amount]: ownershipEmployments.at(type)) // Account for the owner of the buildings.
+		{
+			unitEmployment[job] += amount * percent;
+		}
+	}
+	for (const auto& [job, amount]: unitEmployment) // Track Dependents
+	{
+		unitEmployment[job] = amount / (popTypes.at(job).getDependentRatio().value_or(defaultRatio) + womenJobRate);
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:34]: In order to /0, a mod would have to change the working_adult_ratio in a popType to 0; i.e. 100% dependents and 0% workers.

File: EU4ToVic3Tests/MapperTests/BuildingMapperTests/ProductionMethodMapperTests.cpp [hunk 3]
```diff
@@ -191,4 +197,92 @@ TEST(Mappers_ProductionMethodMapperTests, ApplyRulesUnderTeched)
...
+
+	std::map<std::string, V3::Building> buildingMap;
+	buildingMap.emplace("building_logging_camp", lumberCamp);
+
+	// Load in PM and PMGroup definitions
+	const auto [PMs, PMGroups] = prepPMData();
+
+	const auto& lawPMEstimates = mapper.estimatePMs(countryNoTech, PMs, PMGroups, buildingMap);
+
+	EXPECT_THAT(lawPMEstimates, testing::Contains(testing::Pair("pmg_ownership_capital_building_logging_camp", std::make_tuple(3, 1.0))));
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4ToVic3Tests/MapperTests/BuildingMapperTests/ProductionMethodMapperTests.cpp:258]: 4th pm, not 4th pmg. It's a map of what index to select in each pmg's pm vector.

```
pmg_ownership_capital_building_logging_camp = {
	production_methods = {
		pm_merchant_guilds_building_logging_camp
		pm_privately_owned_building_logging_camp
		pm_publicly_traded_building_logging_camp
		pm_government_run_building_logging_camp // this one
		pm_worker_cooperative_building_logging_camp
	}
}
```

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	for (const auto& [job, amount]: manorHouseRoster) // Peasants create manor house jobs.
+	{
+		unitEmployment[job] += amount;
+	}
+	for (const auto& [job, amount]: unitEmployment) // Track Dependents
+	{
+		unitEmployment[job] = amount / (popTypes.at(job).getDependentRatio().value_or(defaultRatio) + womenJobRate);
+	}
+
+	const double unitPop = std::accumulate(unitEmployment.begin(), unitEmployment.end(), 0.0, [](double sum, const auto& pair) {
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:70]: unitPop is the maximum # of people who could be employed by 1 level of a building, accounting for the jobs created by ownership and urban centers, and all the dependents of those who work in the building.

The only way it could be 0 is if a building does not create any jobs. Which doesn't seem likely, but sure I guess PDX could do that one day, I'll add a check.

I've also renamed unitPop to unitEmploymentPop.

File: EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 5]
```diff
@@ -864,6 +897,63 @@ std::map<std::string, int> V3::EconomyManager::apportionInvestors(const int leve
...
+		{
+			continue;
+		}
+
+		investorWeights[type] = investorData.weight;
+		totalWeight += investorData.weight;
+	}
+	for (const auto& type: investorWeights | std::views::keys)
+	{
+		investorWeights[type] /= totalWeight;
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.cpp:951]: Those would be "national" or "local" investors. Also this function is turning the weights from configurbales/economy/ownership.txt into fractions on a country by country basis. I've refactored a couple things to hopefully paint a clearer picture there.

File: EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingGroups.h [hunk 6]
```diff
@@ -19,6 +19,7 @@ class BuildingGroups
 	[[nodiscard]] std::optional<std::string> tryGetParentName(const std::optional<std::string>& theGroupName) const;
 	[[nodiscard]] std::optional<double> tryGetInfraCost(const std::optional<std::string>& theGroupName) const;
 	[[nodiscard]] std::optional<bool> tryGetIsCapped(const std::optional<std::string>& theGroupName) const;
+	[[nodiscard]] int getUrbanization(const std::optional<std::string>& theGroupName) const;
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingGroups.h:22]: Building::GetBuildingGroup returns a string, so you could just take a string parameter instead of an optional<string> parameter.

File: EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingResources.h
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingResources.h:1]: As I usually ask, please add tests for this class.

File: EU4ToVic3/Source/V3World/EconomyManager/Building/ProductionMethods/ProductionMethod.cpp [hunk 7]
```diff
@@ -58,11 +58,20 @@ void V3::ProductionMethod::registerKeys()
 std::string V3::ProductionMethod::getType(const std::string& typeString)
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Building/ProductionMethods/ProductionMethod.cpp:58]: Some tests for this function would clarify what it's doing.

**Review by Idhrendur — COMMENTED** [review]
I'm about 29% through, will continue in the coming days. For the future, when you've got bits fairly stable, it doesn't hurt to spin them off into smaller PRs so we can review and approve faster.

**Review by Idhrendur — COMMENTED** [review]
86% reviewed

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+	 const double womenJobRate,
+	 const std::map<std::string, double>& estimatedOwnerships,
+	 const std::map<std::string, std::map<std::string, int>>& ownershipEmployments,
+	 const std::map<std::string, PopType>& popTypes,
+	 const std::shared_ptr<SubState>& subState)
+{
+	auto unitEmployment = rgoUnitEmployment;
+	for (const auto& [type, fraction]: estimatedOwnerships)
+	{
+		for (const auto& [job, amount]: ownershipEmployments.at(type)) // Account for the owner of the buildings.
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:27]: Is it at all possible for ownershipEmployments to not have type? If it is possible, you should make sure to handle it in some way or another.

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+		{
+			unitEmployment[job] += amount * fraction;
+		}
+	}
+	for (const auto& [job, amount]: unitEmployment) // Track Dependents
+	{
+		unitEmployment[job] = amount / (popTypes.at(job).getDependentRatio().value_or(defaultRatio) + womenJobRate);
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:34]: You use `popTypes.at(job).getDependentRatio().value_or(defaultRatio) + womenJobRate` in at least a couple places. It'd be more clear if you shoved it into a well-named function in an anonymous namespace at the top of this file.

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+{
+	const double unemployed = subState->getJob("unemployed");
+	if (unemployed > amount)
+	{
+		subState->addJob("unemployed", -amount);
+		return 0;
+	}
+	amount -= unemployed;
+	subState->setJob("unemployed", 0);
+	return amount;
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:108]:
```suggestion
       	subState->setJob("unemployed", 0);
	return amount - unemployed;
```

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+	 const std::map<std::string, double>& subsistenceUnitEmployment,
+	 const double defaultRatio,
+	 const double womenJobRate,
+	 const std::map<std::string, PopType>& popTypes,
+	 const std::shared_ptr<V3::SubState>& subState)
+{
+	// Use peasants as a proxy for subsistence worker presence.
+	if (subState->getJob("peasants") == 0)
+	{
+		// Log(LogLevel::Warning) << "No subsistence workers available.";
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:122]: If this is no longer needed, delete.

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+	{
+		subState->addJob(job, -people * lostSubsistenceLevels);
+	}
+	// When subsistence levels are removed, Manor Houses downsize
+	downsizeManorHouses(lostSubsistenceLevels, defaultRatio, womenJobRate, popTypes, subState);
+
+	// if (const double remainder = amount - lostSubsistenceLevels * unitSubsistencePop; remainder > 0)
+	//{
+	//	Log(LogLevel::Warning) << "Could not find available workers for " << remainder << " jobs."; // Should never happen.
+	// }
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:156]: If no longer needed, delete.

File: EU4ToVic3Tests/V3WorldTests/EconomyManagerTests/DemandTests/MarketJobsTests.cpp [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+	popTypes.emplace("clerks", clerks);
+	popTypes.emplace("capitalists", capitalists);
+	popTypes.emplace("peasants", peasants);
+	popTypes.emplace("farmers", farmers);
+
+	return popTypes;
+}
+} // namespace
+
+TEST(V3World_MarketJobsTests, CreateSubsistenceNotEnoughLand)
```
Comment (Idhrendur) [inline EU4ToVic3Tests/V3WorldTests/EconomyManagerTests/DemandTests/MarketJobsTests.cpp:29]: Tests are more clear when named along the lines of FunctionExpectedResultWhenConditions. I'd guess this to be CreateSubsistenceLeavesUnemployedPopsWhenNotEnoughLand. Wordy, but quite clear. If possible rename this and other test names as appropriate.

**Review by Idhrendur — COMMENTED** [review]
Sorry for the pause, 90% reviewed.

File: EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingResources.h [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+#include "ProductionMethods/ProductionMethodGroup.h"
+
+namespace V3
+{
+class BuildingResources
+{
+  public:
+	BuildingResources() = default;
+
+	void addJobs(const std::string& job, const double amount) { jobs[job] += amount; }
```
[the remark this answers is not in the document]
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Building/BuildingResources.h:13]: So, I discovered in my next PR that since we're hardcoding ports to Siberia and whatnot, there's going to be cases where we do have to build buildings when there isn't enough workforce. In that PR I abandon that absolute limits on building without workforce to allow for the hardcoded buildings, and allowing these estimates to go negative gives more accurate predictions. (It offsets the increase in expected workers elsewhere.)

Because the exact dependent to worker ratio is a little weird at game start, I don't see a benefit to exactly calculating what % a building is going to be filled when that estimate is likely to be off by about a similar amount. So I'd rather just let the negative jobs offset the extra positive ones until proven otherwise.

File: EU4ToVic3/Source/V3World/EconomyManager/Building/ProductionMethods/ProductionMethod.cpp (hunk 7 above)
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Building/ProductionMethods/ProductionMethod.cpp:58]: added a comment and a separate test

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp (hunk 11 above)
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:121]: At this point I was unsure whether or not it was going to be needed. That was determined in the next PR (don't need), and it is deleted there.

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp (hunk 12 above)
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:155]: same as above

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp (hunk 9 above)
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp:34]: iirc Zem does private functions over anonymous namespaces

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp (hunk 8 above)
Comment (gawquon) [reply EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.cpp]: Not currently, but worth logging.

File: EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp [hunk 15]
```diff
@@ -281,6 +281,39 @@ TEST(V3World_CountryTests, PopCounterSumsSubStatePop)
...
+	V3::Country country;
+	std::vector<std::shared_ptr<V3::SubState>> substates;
+
+	EXPECT_EQ(0, country.getPopCount());
+	EXPECT_EQ(0, V3::Country::getPopCount(substates));
+
+	auto sub0 = std::make_shared<V3::SubState>();
+	auto sub1 = std::make_shared<V3::SubState>();
+
+	V3::Pop pop0, pop1, pop2;
```
Comment (Idhrendur) [inline EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp:295]: It would be more clear (and fewer lines) to use the constructor:
```suggestion
    V3:Pop pop0("english", "", "", 1000);
    V3:Pop pop1("german", "", "", 500);
    V3:Pop pop1("english", "", "", 500);
```

File: EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp [hunk 16]
```diff
@@ -333,4 +366,106 @@ TEST(V3World_CountryTests, YearCapFactorHitsCurve)
 	EXPECT_NEAR(0.3, V3::Country::yearCapFactor(date("1650.1.1")), 0.015);
 	EXPECT_NEAR(0.2, V3::Country::yearCapFactor(date("1490.1.1")), 0.01);
 	EXPECT_NEAR(0.15, V3::Country::yearCapFactor(date("1350.1.1")), 0.01);
+}
+
+TEST(V3World_CountryTests, HasAnyOfLawUnlockingPositiveCase)
```
Comment (Idhrendur) [inline EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp:371]:
```suggestion
TEST(V3World_CountryTests, HasAnyOfLawUnlockingIsTrueIfAnyOfLawsArePresent)
```

File: EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp [hunk 17]
```diff
@@ -333,4 +366,106 @@ TEST(V3World_CountryTests, YearCapFactorHitsCurve)
...
+	data.laws.emplace("law_4");
+
+	country.setProcessedData(data);
+
+	EXPECT_TRUE(country.hasAnyOfLawUnlocking({"law_2"}));
+	EXPECT_TRUE(country.hasAnyOfLawUnlocking({"law_1", "law_2"}));
+	EXPECT_TRUE(country.hasAnyOfLawUnlocking({"law_2", "law_3"}));
+}
+
+TEST(V3World_CountryTests, HasAnyOfLawUnlockingNegativeCase)
```
Comment (Idhrendur) [inline EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp:387]:
```suggestion
TEST(V3World_CountryTests, HasAnyOfLawUnlockingFalseIfNoneOfLawsPresent)
```

File: EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp [hunk 18]
```diff
@@ -333,4 +366,106 @@ TEST(V3World_CountryTests, YearCapFactorHitsCurve)
...
+	data.laws.emplace("law_3");
+	data.laws.emplace("law_4");
+
+	country.setProcessedData(data);
+
+	EXPECT_FALSE(country.hasAnyOfLawUnlocking({"law_1"}));
+	EXPECT_FALSE(country.hasAnyOfLawUnlocking({"law_1", "law_5"}));
+}
+
+TEST(V3World_CountryTests, HasAnyOfLawBlockingPositiveCase)
```
Comment (Idhrendur) [inline EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp:402]:
```suggestion
TEST(V3World_CountryTests, HasAnyOfLawBlockingTrueIfAnyOfLawsArePresent)
```

File: EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp [hunk 19]
```diff
@@ -333,4 +366,106 @@ TEST(V3World_CountryTests, YearCapFactorHitsCurve)
...
+	data.laws.emplace("law_4");
+
+	country.setProcessedData(data);
+
+	EXPECT_TRUE(country.hasAnyOfLawBlocking({"law_2"}));
+	EXPECT_TRUE(country.hasAnyOfLawBlocking({"law_1", "law_2"}));
+	EXPECT_TRUE(country.hasAnyOfLawBlocking({"law_2", "law_3"}));
+}
+
+TEST(V3World_CountryTests, HasAnyOfLawBlockingNegativeCase)
```
Comment (Idhrendur) [inline EU4ToVic3Tests/V3WorldTests/PoliticalManagerTests/CountryTests/CountryTests.cpp:418]:
```suggestion
TEST(V3World_CountryTests, HasAnyOfLawBlockingFalseIfNoneOfLawsPresent)
```

File: EU4ToVic3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 20]
```diff
@@ -128,11 +128,37 @@ bool V3::Country::hasAnyOfTech(const std::set<std::string>& techs) const
...
 		return processedData.techs.contains(tech);
 	});
 }
 
+bool V3::Country::hasAnyOfLawUnlocking(const std::set<std::string>& laws) const
+{
+	if (laws.empty())
+	{
+		return true;
+	}
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/PoliticalManager/Country/Country.cpp:141]: Add a test for the empty laws case, especially since that's the difference between this and hasAnyOfLawBlocking().

File: EU4ToVic3/Source/V3World/PoliticalManager/Country/Country.cpp [hunk 21]
```diff
@@ -128,11 +128,37 @@ bool V3::Country::hasAnyOfTech(const std::set<std::string>& techs) const
...
+		return processedData.laws.contains(targetLaw);
+	});
+}
+
+bool V3::Country::hasAnyOfLawBlocking(const std::set<std::string>& laws) const
+{
+	if (laws.empty())
+	{
+		return false;
+	}
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/PoliticalManager/Country/Country.cpp:153]: Ditto

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.h [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+  private:
+	// For a given popType, what % of that popType have jobs?
+	static double calculateWorkerDependencyRatio(const PopType& popType, double defaultRatio, double womenJobRate);
+	// How many people do x number of jobs support?
+	static double getWorkersPlusDependents(double amountOfJobs, const PopType& popType, double defaultRatio, double womenJobRate);
+
+	// Returns the amount of jobs with no unemployed available.
+	static double hireFromUnemployed(double amount, const std::shared_ptr<SubState>& subState);
+	// Returns the level of subsistence buildings downsized.
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketJobs.h:39]:
```suggestion
	// For a given popType, what % of that popType have jobs?
	static double calculateWorkerDependencyRatio(const PopType& popType, double defaultRatio, double womenJobRate);
	// How many people do x number of jobs support?
	static double getWorkersPlusDependents(double amountOfJobs, const PopType& popType, double defaultRatio, double womenJobRate);

	// Returns the amount of jobs with no unemployed available.
	static double hireFromUnemployed(double amount, const std::shared_ptr<SubState>& subState);
	// Returns the level of subsistence buildings downsized.
```

Since these aren't used outside the class, it would be better to put them in an anonymous namespace at the top of the .cpp file and not expose them in the interface.

File: EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketTracker.cpp [hunk 23]
```diff
@@ -0,0 +1,288 @@
...
+	for (const auto& [good, amt]: breakdown)
+	{
+		nameLength = std::max(nameLength, static_cast<int>(good.length()));
+		percentLength = std::max(percentLength, static_cast<int>(std::to_string(amt).length()));
+	}
+
+	out << std::setprecision(3);
+	out << std::endl;
+	out << std::setfill('-') << std::setw(nameLength + 2) << "" << std::setw(percentLength + 2) << "" << std::endl;
+	out << std::setfill(' ');
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/Demand/MarketTracker.cpp:191]: FWIW, this becomes easier (and easier to read) by pulling in fmt::format (to be replaced by std::format when our runners can take it).

File: EU4ToVic3Tests/V3WorldTests/EconomyManagerTests/DemandTests/MarketTrackerTests.cpp
Comment (Idhrendur) [inline EU4ToVic3Tests/V3WorldTests/EconomyManagerTests/DemandTests/MarketTrackerTests.cpp:1]: Remember to write tests here, as much as possible.

File: EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.h [hunk 24]
```diff
@@ -89,34 +94,44 @@ class EconomyManager
 	[[nodiscard]] double calculateStateTraitMultiplier(const std::shared_ptr<SubState>& subState) const;
 	[[nodiscard]] double getDensityFactor(double perCapitaDev) const;
 
+	void establishBureaucracy(const std::shared_ptr<Country>& country, const std::map<std::string, Law>& lawsMap, const Vic3DefinesLoader& defines) const;
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.h:97]:
```suggestion
	void establishBureaucracy(const Country& country, const std::map<std::string, Law>& lawsMap, const Vic3DefinesLoader& defines) const;
```

You're not taking (shared) ownership of the country, so just take a reference to it.

File: EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.h [hunk 25]
```diff
@@ -89,34 +94,44 @@ class EconomyManager
...
+	void hardcodePorts(const std::shared_ptr<Country>& country) const;
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.h:98]:
```suggestion
	void hardcodePorts(const Country& country) const;
```

Ditto.

File: EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.cpp [hunk 26]
```diff
@@ -72,54 +73,75 @@ void V3::EconomyManager::establishBureaucracy(const PoliticalManager& politicalM
...
+void V3::EconomyManager::integrateHardcodedBuildings(const std::shared_ptr<Country>& country,
+	 double defaultRatio,
+	 const std::map<std::string, std::tuple<int, double>>& estimatedPMs,
+	 const std::map<std::string, Law>& lawsMap,
+	 const std::map<std::string, PopType>& popTypes,
+	 MarketTracker& market) const
+{
+	for (const auto& subState: country->getSubStates())
+	{
+		for (const auto& building: subState->getBuildings())
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.cpp:124]: Are you getting a shared_ptr to a building? Maybe see if you can dereference it immediately in the loop construct.

File: EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.h [hunk 27]
```diff
@@ -89,34 +94,44 @@ class EconomyManager
...
+	void integrateHardcodedBuildings(const std::shared_ptr<Country>& country,
```
Comment (Idhrendur) [inline EU4ToVic3/Source/V3World/EconomyManager/EconomyManager.h:99]:
```suggestion
	void integrateHardcodedBuildings(const Country& country,
```
Ditto

**Review by Idhrendur — APPROVED** [review]

File: EU4ToVic3Tests/V3WorldTests/EconomyManagerTests/DemandTests/MarketTrackerTests.cpp
Comment (gawquon) [reply EU4ToVic3Tests/V3WorldTests/EconomyManagerTests/DemandTests/MarketTrackerTests.cpp:1]: Addressed in next PR

**Review by Zemurin — APPROVED** [review]
